docs/life-of-a-request.md

```diff
@@ -0,0 +1,200 @@
+
+Static files include javascript, css, static assets (like emoji, avatars),
+and user uploads (if stored locally and not on S3).
+
+```
+/home/zulip/prod-static
+```
+
+## Nginx routes other requests between tornado and django
+
+All our connected clients hold open long-polling connections so that
+they can recieve events (messages, presence notifications, and so on) in
+real-time. Events are served by Zulip's `tornado` application.
+
+Nearly every other kind of request is served by the `zerver` Django
+application.
+
+# Django routes the request to a view in urls.py files
+
+There are various urls.py files throughout the server codebase, and you
+can read about it more in [the directory structure doc](http://zulip.readthedocs.io/en/latest/directory-structure.html).
+
+The main Zulip Django app is `zerver`. The routes are found in
+```
+zproject/urls.py
+zproject/legacy_urls.py
+```
+
+There are HTML-serving, REST API, legacy, and webhook url patterns.
+
+## Views serving HTML are internationalized by server path
+
+If you look in [zproject/urls.py](https://github.com/zulip/zulip/blob/master/zproject/urls.py), you'll see something called
+`i18n_urls`. These urls show up in the address bar of the browser, and
+serve HTML.
+
+For example, the `/hello` page (preview [here](https://zulip.com/hello/))
+gets translated in Chinese at `zh-cn/hello/` (preview [here](https://zulip.com/zh-cn/hello/)).
+
+Note the `zh-cn` prefix--that url pattern gets added by `i18n_patterns`.
+
+## API endpoints use REST
+
+With the exception of Webhooks (which we do not control), and legacy
+endpoints, and logged-out endpoints, Zulip uses REST for its API. This
+means that we use:
+
+* POST for creating something new where you don't have a unique ID
+* PUT for creating or completely replacing something for which you have
+  a unique ID
+* DELETE for deleting something
+* PATCH for updating or editing attributes of something.
+* GET to get something (read-only)
+* HEAD to check the existence of something to GET, without getting it;
+  useful to check a link without downloading a potentially large link
+* OPTIONS (handled automatically, see more below)
+* POST for everything else
+
+Of these, PUT, DELETE, HEAD, OPTIONS, and GET are *idempotent*, which
+means that you can send the request multiple times and get the same
+state on the server. POST is not idempotent--if I send a message
+multiple times, I will get multiple message. PATCH is special--it can be
+idempotent, and we like it if you write API endpoints where it is
+idempotent, as much as possible.
+
+### PUT vs POST vs PATCH
+
+The difference between these verbs is subtle at first glance, but
+important.
+
+Let's say you have something (the HTTP RFC calls them "resources") like
+a realm emoji. An emoji has two attributes: a name and a url for the
+emoji image. Let's pretend there's also an emoji alternative name
+attribute.
+
+In this case, we should avoid POST, since we have a unique identifier:
```

Comment:
This may be unintuitive for some readers given that the example used elsewhere (creating a user) uses `POST` even though we have a unique identifier: the user's email.